blueprints/codes/__init__.py

```diff
@@ -0,0 +1,56 @@
+"""Module for the abstract base class Equation."""
+from abc import ABC, abstractmethod
+
+
+class Equation(ABC):
```

Comment:
A nice start, but as this is a quite essential class of our library I have An alternative proposol for the solution architecture of this. --> It is a bit hard to explain in this comment, so maybe we should have a small brainstorm session with a small amount of people:

### Proposal:

It is still a a 'mindgame' for me, but look at this idea, we can take this as a start:

Let's distinguise a 'variable' class and an "equation" class. 

The variable class is essentially a float in usage with extra attributes: a variable name, and a symbol and maybe even a description (because there are many "alphas" in the EC)

The equation class is a composition of multiple variables, it actually should act like a callable (function) because I think that is the most intuitive way of using it for a user the result of an equation: a variable. The equation class itself holds a symbolic reprensentation  (if you want to use it in a report) of the equation and things like: the source document, equation label etc.

The resulting variable stores the intermediate values of the variables, the actual equation object, the source, When we design the 'variable' as a "tree" [Composite design pattern]  the 'variable' should be able to keep track of a 'equation-pipe-line' (parent-child dependecies). By using recursion you should be able to reconstruct the complete intermediate results. I'm not sure yet if it is a problem if using a variable multiple times in different equations, maybe the child parent relationsship solves this right away.

It is a bit complex, but if we can do this correctly it would be a amazing feature that can create a complete report of a calculation without the user having any trouble doing this.

But as I said: this is only something that is spinning around in my mind. Let me know what you think. Happy to discuss this to take it to a higher level :-)

Reply:
Hey @bro-wi, I can imagine that it is hard to put this idea on words. I'm up for a small brainstorm somewhere this week (do you guys prefer business hours or nights?). I'm pretty much in line with you when you say we got to get this one right. :-)

Reply:
I'd prefer business hours, my nights are terrible at the moment (wait for it) :-p

some timeslots this week:

today: until 16.00
Wednesday: 9.00-14.00
Thursday: 9.00-14.00

Reply:
11.30 today?
https://teams.microsoft.com/l/meetup-join/19%3ameeting_NzEzNGI0YTEtOGM0Yy00ZTgyLWFmOTAtYzgyZTdhNDA2ZDcy%40thread.v2/0?context=%7b%22Tid%22%3a%22bf5f4046-a1dc-4119-aa8a-bfb9fbf46271%22%2c%22Oid%22%3a%22b1980a8f-5dff-4ec7-b9f5-d95ea2af341d%22%7d

Reply:
As discussed, I tried prototyping some stuff, 
 this is what I came up with:

# First of all:
IDEA: can we include the design decisions in the Issues? for example: why do we choose to use "Equation" as a name and not "Formula", the solution architecture and interface we choose to use etc. etc. 

I think it is good to have a discussions on those matters prior to coding, now we are discussing in a piece of code that is already done: it makes it hard to move directions.



# Proposals:
I know these proposals are incomplete and absolutely not on a polish level, but I quickly wanted to test de functionality without worrying about variable_names, structures etc --> this can be done when we choose the actual implementation.

## proposal 1: close to the curret implementation
In this proposal I kept it as close to the current impelemtation  with the extension that you are able to use a formula as input value.

Benefits:
- return types of methods are quite flexible to implement for each function
- design is quite simple

Downsides:
- using the result is quite counter intuitive (it is not a float or something: you must use class.result to use the actual result
- I don't like te use a class that is actually a function (But that is a matter of taste)

``` 
# ? Proposal 1, closest to current design
# TODO: Design choice: is it an Equation, Expression, Formula , Identity --> https://www.mathswithgraham.org.uk/?p=2873
class Formula:
    equation_label = "6.1"
    source_document = "EC15"

    def __init__(self, *, a: float | Self, b: float | Self, c: float | Self) -> None:
        self.a = a.result() if isinstance(a, Formula) else a
        self.b = b.result() if isinstance(b, Formula) else b
        self.c = c.result() if isinstance(c, Formula) else c
        self.vars = {"a": a, "b": b, "c": c}

    def result(self) -> float:
        return self.a + self.b + self.c

    def detailed_result(self) -> dict[str, Any]:
        calculation_steps = {
            "formula": f"X = a + b + c = {self.result()}    ({self.equation_label})"
        }

        for var, value in self.vars.items():
            if isinstance(value, Formula):
                calculation_steps[var] = value.detailed_result()
            else:
                calculation_steps[var] = value

        # optionally extend these results with specific intermediate results

        return calculation_steps


# when using this structure you can retrieve the complete "function pipeline"
x = Formula(a=1, b=3, c=3)
y = Formula(a=1, b=x, c=1)
y.equation_label = "6.2"  # changing the label for better traceable results
z = Formula(a=y, b=x, c=4)
z.equation_label = "6.3"

for formula in (x, y, z):
    print(f"detailed answer for {formula.equation_label}")
    print(json.dumps(formula.detailed_result(), indent=4))
    print(10 * "=")
```

## proposal 2: the function object acts like a float
In this proposal the function is acting like a float and the result can thus be used directly (i.o using class.result)

Benefits:
- design is quite simple
- usage of the result is intuitive and easy

Downsides:
- I don't like te use a class that is actually a function (But that is a matter of taste), but in this case it feels more like a function because you don't need the ".result" notation.
- when multiple return types are possible we have to create multiple objects (not sure if this actually happens)

``` 
# ? Proposal 2, Class where the formula is acting like a float so it can be used directly without using the .result() method.
class FormulaAsFloat(float):
    equation_label = "6.1"
    source_document = "EC15"

    def __new__(cls, *, a, b, c) -> Self:
        return float.__new__(cls, a + b + c)

    def __init__(self, *, a: float | Self, b: float | Self, c: float | Self) -> None:
        self.a = a
        self.b = b
        self.c = c
        self.vars = {"a": a, "b": b, "c": c}

    def detailed_result(self) -> dict[str, Any]:
        calculation_steps = {
            "formula": f"X = a + b + c = {self}    ({self.equation_label})"
        }

        for var, value in self.vars.items():
            if isinstance(value, FormulaAsFloat):
                calculation_steps[var] = value.detailed_result()
            else:
                calculation_steps[var] = value

        # Optionally extend these results with specific intermediate results

        return calculation_steps


# when using this structure you can retrieve the complete "function pipeline"
x = FormulaAsFloat(a=1, b=3, c=3)
y = FormulaAsFloat(a=1, b=x, c=1)
y.equation_label = "6.2"  # changing the label for better traceable results
z = FormulaAsFloat(a=y, b=x, c=4)
z.equation_label = "6.3"

for formula in (x, y, z):
    print(f"detailed answer for {formula.equation_label}")
    print(f"answer={formula}") #here you see that it acts like a float, we could even implement a __str__ method with units for the answer
    print(json.dumps(formula.detailed_result(), indent=4))
    print(10 * "=")
```


## proposal 3: using a function, but with function attributes

This one is not worked out properly yet, and i'm not really sure if this is what we want: but maybe it inspires someone else.
This could be an idea: It could work if we just want to create a function that outputs something and the function has function attributes with the equation_label and document--> on the "function-pipeline" part this takes a bit more work but generally the idea could be something like this:


``` 
class FunctionOutput(float):
    def __new__(cls, value, **kwargs) -> Self:
        return float.__new__(cls, value)

    def __init__(self, value, *, formula, detailed_results):
        self.formula = formula
        self.detailed_results = detailed_results


def formula_test(*, a, b, c):
    formula_test.label = "6.10"
    formula_test.source_document = "EC512540"

    return FunctionOutput(a + b + c, formula=formula_test, detailed_results="this")


result = formula_test(a=1, b=2, c=3)
print(result)
print(result.formula.label)
```

Just let me know what everybody else think, if these designs are undesirable we just ditch it, but In that case at least we have investigated multiple options.

Reply:
@bro-wi, great work on developing these proposals! You truly think outside the box :-)

I like the solutions presented in the first two proposals. I tried to improve upon them by adding a clear interface to all subclasses of `Formula`, using an Abstract Base Class (ABC). In this new design, the result method is specifically used to separate the logic when using the __new__ method, which helps in maintaining a clear structure and readability. What do you guys think @Blueprints-org/maintainers? I believe we're almost there.

Pros:
- Consistency: Ensures a standard structure for all formula subclasses.
- Separation of Logic: Distinctly separates calculation logic (result) from object instantiation (__new__).
- Flexible Implementation: Allows different calculation logic for various formulas.
- Intuitive Usage: Operates like a float, making it user-friendly. No need to use .result().
- Detailed Documentation: Supports comprehensive representation and documentation of formulas.

Cons:

- Complex for Simple Tasks: May be overly complex for basic calculations.
- Learning Curve: Could require additional time for new users to learn.
- Limited to Float Results: Primarily designed for float outputs, might need adjustments for other return types.


```python
from abc import ABC, abstractmethod
from typing import Any


class Formula(float, ABC):
    def __new__(cls, *args, **kwargs) -> 'Formula':
        result = cls.result(*args, **kwargs)
        return float.__new__(cls, result)

    @staticmethod
    @abstractmethod
    def result(*args, **kwargs) -> float:
        pass

    @abstractmethod
    def view_latex(self) -> str:
        pass

    @abstractmethod
    def source_document(self) -> str:
        pass

    @abstractmethod
    def equation_label(self) -> str:
        pass


class Form5Dot2SpecificFormula(Formula):
    equation_label = "5.2"
    source_document = "EC2 1992-1-1"

    def __init__(self, a: float, b: float, additional_param: float) -> None:
        self.a = a
        self.b = b
        self.additional_param = additional_param
        self.vars = {"a": a, "b": b, "additional_param": additional_param}

    @staticmethod
    def result(a: float, b: float, additional_param: float) -> float:
        """Main logica calculation to be performed."""
        return a + b + additional_param

    def view_latex(self) -> str:
        """One of the possible available methods to represent the formula."""
        return f"X = {self.a} + {self.b} + {self.additional_param} = {self}"


class Form7Dot8AnotherFormula(Formula):
    equation_label = "7.8"
    source_document = "EC2 1992-1-1"

    def __init__(self, q: float, p: float, r: float) -> None:
        self.q = q
        self.p = p
        self.r = r
        self.vars = {"q": q, "p": p, "r": r}

    @staticmethod
    def result(q: float, p: float, r: float) -> float:
        """Main logica calculation to be performed."""
        return (q + p) / r

    def view_latex(self) -> str:
        """One of the possible available methods to represent the formula."""
        return f"X = ({self.q} + {self.p}) / {self.r} = {self}"


# Example usage
form_5_2 = Form5Dot2SpecificFormula(a=1, b=3, additional_param=2)
form_7_8 = Form7Dot8AnotherFormula(q=1, p=form_5_2, r=2)
print(f"Result of form_5_2: {form_5_2}")  # Output will be the float result of the calculation
print(f"Result of form_7_8: {form_7_8}")  # Output will be the float result of the calculation
print(f"LaTeX representation of form_5_2: {form_5_2.view_latex()}")
print(f"LaTeX representation of form_7_8: {form_7_8.view_latex()}")

# Result of form_5_2: 6.0
# Result of form_7_8: 3.5
# LaTeX representation of form_5_2: X = 1 + 3 + 2 = 6.0
# LaTeX representation of form_7_8: X = (1 + 6.0) / 2 = 3.5

```

P.S>I'm aware I'm missing the chaining of detailed_results which I liked on your second proposal. I'm so sad a day has only 24 hours :-(

Reply:
Haha, don't be sad!
I like your additions, and totally agree about separating initiation vs calculation.

Two suggestions:
1. I would suggest to change the ".result()" method into ".evaluate()" as your are performing an calculation. It feels more like an 'action': but it is a matter of taste.
2. we could consider to introduce a "Variable" class (also a float) with a "view_ latex" method (maybe even a unit?)  --> Then you can use that directly in the 'view_latex' method in the formula. That is possibly useful when multiple variables are used in multiple formula's (e.g. rho etc.). Maybe that is overkill for now...


I also agree on the cons, it is a bit complex for developing, (although when the ABC is already there, we already have taken care of that aspect). So we have to weigh if the pros against the cons.

About the the 'only-float' results. It is true, for some types it is doable (when we change the order of succesion [so the __new__ a the subclass) but:
if you would like to create a bool type it is not even possible: the boolean is not subclassable because it is a Singleton. https://mail.python.org/pipermail/python-dev/2002-March/020822.html

I cant really overview how often another return type (other than float) would occur... who can jump in on that?

Reply:
I agree, evaluate() instead of result() maybe makes more sense. Glad we're coming to a solution. I will work out the details and push them to this PR again. Hopefully we can start soon writing the Formulas :-)

Reply:
Done ;-)